File: apps/ios/App/Billing/PaywallSheet.swift
import StoreKit
import SwiftUI
import TrackEvolutionKit

/// The paywall (NS-32 rule 5; App Store guideline 3.1.2): what Pro is, what it
/// costs per term in the store's own localised price, how it renews, Restore
/// Purchases, and the privacy policy and terms — every one of which review looks
/// for. A sheet, not a disabled control: it names the price and lets you buy on
/// the spot.
///
/// Presented from a *button* — the record screen's Start, Settings' Subscribe —
/// never from a screen's root. Those screens already carry
/// a modal, and two presentations on one view is a documented way to wedge
/// SwiftUI here (`apps/ios/README.md`).
///
/// Nothing in this file decides tier. It reads `auth.entitlement`, which the
/// server wrote, and asks `StoreController` to sell.
struct PaywallSheet: View {
    /// What brought the sheet up, for the headline.
    enum Context {
        case general
        case record
    }

    var context: Context = .general

    @Environment(StoreController.self) private var store
    @Environment(AuthController.self) private var auth
    @Environment(\.dismiss) private var dismiss
    @State private var selectedId: String?

    static let privacyURL = URL(string: "https://docs.trackevolution.app/docs/privacy.html")!
    static let termsURL = URL(string: "https://docs.trackevolution.app/docs/terms.html")!

    private var isPro: Bool { Entitlement.isPro(auth.entitlement) }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: TESpacing.gridGap) {
                    header
                    tiersCard
                    if isPro {
                        proCard
                    } else {
                        purchaseCard
                    }
                    legal
                }
                .padding(TESpacing.pageGutter)
            }
            .background(Color(.bgPage))
            .navigationTitle("Track Evolution Pro")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarTrailing) {
                    Button("Done") { dismiss() }
                        .accessibilityIdentifier("paywallDone")
                }
            }
        }
        .task {
            if store.productsState != .loaded { await store.loadProducts() }
            if selectedId == nil { selectedId = store.products.first?.id }
        }
        .onChange(of: store.products) { _, products in
            if selectedId == nil { selectedId = products.first?.id }
        }
        .onChange(of: isPro) { _, pro in
            if pro { Haptics.confirm() }
        }
    }

    // MARK: - Copy

    private var header: some View {
        HStack(alignment: .top, spacing: 14) {
            BrandMark(size: 44)
            VStack(alignment: .leading, spacing: 4) {
                Text(headline)
                    .teStyle(.h2)
                    .foregroundStyle(Color(.textStrong))
                Text("Free is the logbook. Pro is the analysis.")
                    .teStyle(.sm)
                    .foregroundStyle(Color(.textMuted))
            }
        }
    }

    private var headline: String {
        switch context {
        case .general: "Turn lap times into analysis"
        case .record: "Recording laps is a Pro feature"
        }
    }

    private var tiersCard: some View {
        TECard {
            VStack(alignment: .leading, spacing: 14) {
                tier("Free", items: [
                    "Tracks, events, sessions and lap times — unlimited",
                    "Lap times out of PDR and GoPro video, with the racing line",
                    "Best laps, progress charts and consistency",
                    "Public share pages and leaderboards",
                    "Works offline, on every device"
                ])
                Divider().overlay(Color(.borderHairline))
                tier("Pro", accent: true, items: [
                    "GPS lap recorder with live timing and predictive delta",
                    "Channel graphs: speed, throttle, brake, steering, RPM",
                    "Sector splits, theoretical best and two-lap compare",
                    "Garage consumables: pad, tire and fluid wear"
                ])
            }
        }
    }

    private func tier(_ name: String, accent: Bool = false, items: [String]) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(name)
                .teStyle(.eyebrow)
                .foregroundStyle(accent ? Color(.accentInk) : Color(.textMuted))
            ForEach(items, id: \.self) { item in
                HStack(alignment: .firstTextBaseline, spacing: 8) {
                    Image(systemName: accent ? "checkmark.circle.fill" : "circle")
                        .teStyle(.xs)
                        .foregroundStyle(accent ? Color(.accentInk) : Color(.textFaint))
                    Text(item)
                        .teStyle(.sm)
                        .foregroundStyle(Color(.textBody))
                }
            }
        }
    }

    // MARK: - Buying

    @ViewBuilder
    private var purchaseCard: some View {
        TECard {
            VStack(alignment: .leading, spacing: 12) {
                switch store.productsState {
                case .idle, .loading:
                    HStack(spacing: 10) {
                        ProgressView()
                        Text("Loading prices from the App Store…")
                            .teStyle(.sm)
                            .foregroundStyle(Color(.textMuted))
                    }
                case .failed(let message):
                    TEErrorBanner(message: message)
                    Button("Try again") { Task { await store.loadProducts() } }
                        .buttonStyle(TEButtonStyle(kind: .quiet))
                case .loaded:
                    ForEach(store.products, id: \.id) { product in
                        productRow(product)
                    }
                    subscribeButton
                }

                Button(store.isRestoring ? "Restoring…" : "Restore Purchases") {
                    Task { await store.restore() }
                }
                .buttonStyle(TEButtonStyle(kind: .quiet))
                .disabled(store.isRestoring || store.isPurchasing)
                .accessibilityIdentifier("paywallRestore")

                if let error = store.error {
                    TEErrorBanner(message: error)
                }
                if let notice = store.notice {
                    Text(notice)
                        .teStyle(.xs)
                        .foregroundStyle(Color(.textMuted))
                }

                Text("""
                    Payment is charged to your Apple ID account when you confirm the purchase. The \
                    subscription renews automatically at the same price and term unless you cancel \
                    at least 24 hours before the current period ends. Manage or cancel it any time \
                    in your App Store account settings.
                    """)
                    .teStyle(.xxs)
                    .foregroundStyle(Color(.textFaint))
            }
        }
    }

    private func productRow(_ product: Product) -> some View {
        let selected = product.id == selectedId
        return Button {
            selectedId = product.id
            Haptics.select()
        } label: {
            HStack(alignment: .center, spacing: 12) {
                VStack(alignment: .leading, spacing: 2) {
                    Text(product.displayName)
                        .teStyle(.bodyStrong)
                        .foregroundStyle(Color(.textStrong))
                    Text(Self.termLine(product, introEligible: store.introEligible.contains(product.id)))
                        .teStyle(.xs)
                        .foregroundStyle(Color(.textMuted))
                }
                Spacer(minLength: 8)
                Text(product.displayPrice)
                    .teStyle(.lapTime)
                    .foregroundStyle(Color(.textStrong))
                Image(systemName: selected ? "largecircle.fill.circle" : "circle")
                    .foregroundStyle(selected ? Color(.accentInk) : Color(.textFaint))
            }
            .padding(12)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Color(.surfaceRaised), in: .rect(cornerRadius: TERadius.md))
            .overlay(
                RoundedRectangle(cornerRadius: TERadius.md)
                    .strokeBorder(selected ? Color(.accent) : Color(.borderHairline), lineWidth: selected ? 2 : 1)
            )
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("paywallProduct-\(product.id)")
        .accessibilityAddTraits(selected ? .isSelected : [])
    }

    private var selectedProduct: Product? {
        store.products.first { $0.id == selectedId }
    }

    @ViewBuilder
    private var subscribeButton: some View {
        if let product = selectedProduct {
            Button(store.isPurchasing ? "Purchasing…" : "Subscribe for \(product.displayPrice) \(Self.perTerm(product))") {
                Task { await store.purchase(product) }
            }
            .buttonStyle(TEButtonStyle(kind: .accent))
            .disabled(store.isPurchasing || store.isRestoring)
            .accessibilityIdentifier("paywallSubscribe")
        }
    }

    // MARK: - Already Pro

    private var proCard: some View {
        TECard {
            VStack(alignment: .leading, spacing: 10) {
                Text("You're Pro")
                    .teStyle(.h3)
                    .foregroundStyle(Color(.textStrong))
                Text(Entitlement.entitlementSummary(auth.entitlement, fmtDate: Self.fmtDate))
                    .teStyle(.sm)
                    .foregroundStyle(Color(.textMuted))
                    .accessibilityIdentifier("paywallSummary")
                switch auth.entitlement?.source {
                case .legacy:
                    Text("You bought the app before subscriptions — Pro is yours for life. Thank you.")
                        .teStyle(.xs)
                        .foregroundStyle(Color(.textFaint))
                case .apple:
                    Button("Manage subscription") { Task { await store.showManageSubscriptions() } }
                        .buttonStyle(TEButtonStyle(kind: .quiet))
                case .google:
                    if let url = Entitlement.manageUrl(auth.entitlement) {
                        Link("Manage on Google Play", destination: url)
                            .buttonStyle(TEButtonStyle(kind: .quiet))
                    }
                case nil:
                    EmptyView()
                }
                if let error = store.error {
                    TEErrorBanner(message: error)
                }
            }
        }
    }

    // MARK: - Legal (guideline 3.1.2)

    private var legal: some View {
        HStack(spacing: 18) {
            Link("Privacy policy", destination: Self.privacyURL)
                .accessibilityIdentifier("paywallPrivacy")
            Link("Terms of use", destination: Self.termsURL)
                .accessibilityIdentifier("paywallTerms")
        }
        .teStyle(.xs)
        .foregroundStyle(Color(.accentInk))
        .frame(maxWidth: .infinity)
    }

    // MARK: - Formatting

    /// "per month", "per year" — the term as the store defines it.
    static func perTerm(_ product: Product) -> String {
        guard let period = product.subscription?.subscriptionPeriod else { return "" }
        return "per \(periodLabel(period))"
    }

    /// "Billed monthly" / "Billed yearly", led by the trial when this Apple ID is
    /// eligible: "2 weeks free, then billed monthly".
    static func termLine(_ product: Product, introEligible: Bool) -> String {
        guard let subscription = product.subscription else { return "" }
        let billed = "billed \(adverb(subscription.subscriptionPeriod))"
        if introEligible, let offer = subscription.introductoryOffer, offer.paymentMode == .freeTrial {
            return "\(periodLabel(offer.period, plural: true)) free, then \(billed)"
        }
        return billed.prefix(1).uppercased() + billed.dropFirst()
    }

    /// "month", "2 weeks", "year".
    static func periodLabel(_ period: Product.SubscriptionPeriod, plural: Bool = false) -> String {
        let unit: String = switch period.unit {
        case .day: "day"
        case .week: "week"
        case .month: "month"
        case .year: "year"
        @unknown default: "period"
        }
        return period.value == 1 ? (plural ? "1 \(unit)" : unit) : "\(period.value) \(unit)s"
    }

    private static func adverb(_ period: Product.SubscriptionPeriod) -> String {
        switch (period.unit, period.value) {
        case (.month, 1): "monthly"
        case (.year, 1): "yearly"
        case (.week, 1): "weekly"
        case (.day, 1): "daily"
        default: "every \(periodLabel(period))"
        }
    }

    static func fmtDate(_ ms: Int) -> String {
        Date(timeIntervalSince1970: Double(ms) / 1000).formatted(date: .abbreviated, time: .omitted)
    }
}

File: apps/android/app/src/main/kotlin/app/trackevolution/billing/PaywallSheet.kt
package app.trackevolution.billing

import android.app.Activity
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import app.trackevolution.ui.TEErrorBanner
import app.trackevolution.ui.theme.TrackCard
import app.trackevolution.ui.theme.TrackTheme

/**
 * The paywall (NS-32 rule 5): a sheet, never a disabled button. It names the
 * price and the term from Play's own `ProductDetails` (so it is localised and
 * agrees with what Play charges), carries both legal links, and has a Restore.
 *
 * The copy draws the tier line the spec fixes: **Free is the logbook, Pro is the
 * analysis.** Nothing anyone already has is behind this sheet.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PaywallSheet(
    billing: BillingController,
    onDismiss: () -> Unit,
    onOpenLink: (String) -> Unit,
) {
    val colors = TrackTheme.colors
    val type = TrackTheme.typography
    val state by billing.state.collectAsState()
    val activity = LocalContext.current as? Activity

    LaunchedEffect(Unit) { if (state.plans.isEmpty()) billing.refreshProducts() }

    // Yearly pre-selected: it is the better deal, and the card says by how much.
    var selectedPlan by rememberSaveable { mutableStateOf(BillingProducts.YEARLY) }
    val plans = state.plans
    val monthly = plans.firstOrNull { it.basePlanId == BillingProducts.MONTHLY }
    val yearly = plans.firstOrNull { it.basePlanId == BillingProducts.YEARLY }
    val selected = plans.firstOrNull { it.basePlanId == selectedPlan } ?: plans.firstOrNull()
    val savings = PaywallPlans.savingsPercent(monthly, yearly)

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        containerColor = colors.surfaceRaised,
        modifier = Modifier.testTag("paywall"),
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 20.dp)
                .padding(bottom = 28.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text("TRACK EVOLUTION PRO", style = type.eyebrow, color = colors.accentInk)
            Text("Turn lap times into analysis", style = type.h1, color = colors.textStrong)
            Text(
                "The logbook is free — tracks, events, sessions, lap times, best laps, progress " +
                    "charts, sharing and leaderboards, with no limit, and lap times out of Corvette " +
                    "PDR and GoPro video. Pro is the analysis: the GPS lap recorder with live timing, " +
                    "the per-lap channel graphs and sector splits inside those same videos, and the " +
                    "garage's consumable wear tracking.",
                style = type.sm,
                color = colors.textMuted,
            )

            when {
                state.loadingPlans -> Row(verticalAlignment = Alignment.CenterVertically) {
                    CircularProgressIndicator(modifier = Modifier.height(18.dp), color = colors.accent)
                    Text(
                        "Fetching prices from Google Play…",
                        style = type.sm,
                        color = colors.textMuted,
                        modifier = Modifier.padding(start = 10.dp),
                    )
                }

                plans.isEmpty() -> {
                    Text(
                        state.unavailable ?: "Prices aren't available right now.",
                        style = type.sm,
                        color = colors.textMuted,
                    )
                    TextButton(onClick = billing::refreshProducts) {
                        Text("Try again", style = type.bodyStrong, color = colors.accentInk)
                    }
                }

                else -> Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    plans.forEach { plan ->
                        PlanCard(
                            plan = plan,
                            selected = plan.basePlanId == selected?.basePlanId,
                            badge = if (plan.basePlanId == BillingProducts.YEARLY && savings != null) "Save $savings%" else null,
                            onClick = { selectedPlan = plan.basePlanId },
                            modifier = Modifier.weight(1f),
                        )
                    }
                }
            }

            Button(
                onClick = { if (selected != null && activity != null) billing.buy(activity, selected) },
                enabled = selected != null && activity != null && !state.busy,
                modifier = Modifier.fillMaxWidth().testTag("paywallSubscribe"),
                colors = ButtonDefaults.buttonColors(
                    containerColor = colors.accent,
                    contentColor = colors.accentContrast,
                ),
            ) {
                Text(
                    when {
                        state.busy -> "Opening Google Play…"
                        selected == null -> "Subscribe"
                        selected.trial != null -> "Start ${selected.trial} — then ${selected.price} / ${selected.period}"
                        else -> "Subscribe — ${selected.price} / ${selected.period}"
                    },
                    style = type.bodyStrong,
                )
            }

            Text(
                "Renews automatically until cancelled in Google Play; cancel any time and keep Pro to " +
                    "the end of the period. A lapse never touches your logbook — every lap you've " +
                    "recorded or entered stays, and nothing you save is ever rejected.",
                style = type.xs,
                color = colors.textFaint,
            )

            TEErrorBanner(state.message)

            Row(
                horizontalArrangement = Arrangement.spacedBy(16.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                TextButton(onClick = billing::restore, enabled = !state.busy, modifier = Modifier.testTag("paywallRestore")) {
                    Text("Restore purchases", style = type.sm, color = colors.textMuted)
                }
                TextButton(onClick = onDismiss) {
                    Text("Not now", style = type.sm, color = colors.textMuted)
                }
            }

            Row(horizontalArrangement = Arrangement.spacedBy(16.dp)) {
                LegalLink("Privacy policy", LegalLinks.PRIVACY, onOpenLink)
                LegalLink("Terms of use", LegalLinks.TERMS, onOpenLink)
            }
        }
    }
}

@Composable
private fun PlanCard(
    plan: Plan,
    selected: Boolean,
    badge: String?,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = TrackTheme.colors
    val type = TrackTheme.typography
    TrackCard(
        modifier = modifier
            .clickable(onClick = onClick)
            .semantics {
                contentDescription = buildString {
                    append(plan.title)
                    append(", ")
                    append(plan.price)
                    append(" per ")
                    append(plan.period)
                    plan.trial?.let { append(", $it") }
                    if (selected) append(", selected")
                }
            }
            .testTag("plan-${plan.basePlanId}"),
        border = if (selected) colors.accent else colors.borderHairline,
        contentPadding = 12.dp,
    ) {
        Text(plan.title, style = type.h3, color = if (selected) colors.accentInk else colors.textStrong)
        Text(plan.price, style = type.h2, color = colors.textStrong, maxLines = 1)
        Text("per ${plan.period}", style = type.xs, color = colors.textMuted)
        plan.trial?.let { Text(it, style = type.xxs, color = colors.accentInk, modifier = Modifier.padding(top = 4.dp)) }
        badge?.let { Text(it, style = type.xxs, color = colors.textFaint, modifier = Modifier.padding(top = 4.dp)) }
    }
}

@Composable
private fun LegalLink(label: String, url: String, onOpenLink: (String) -> Unit) {
    Text(
        "$label ↗",
        style = TrackTheme.typography.xs.copy(textDecoration = TextDecoration.Underline),
        color = TrackTheme.colors.textMuted,
        modifier = Modifier.clickable { onOpenLink(url) },
    )
}
